lib/tapioca/autoload_tracker.rb

```diff
@@ -0,0 +1,32 @@
+# typed: true
+# frozen_string_literal: true
+
+module Tapioca
+  module AutoloadTracker
+    extend T::Sig
+
+    @constant_names_registered_for_autoload = T.let([], T::Array[String])
+
+    def autoload(module_name, filename)
+      AutoloadTracker.register("#{self}::#{module_name}")
+      super
+    end
+
+    sig { void }
+    def self.eager_load_all!
+      until @constant_names_registered_for_autoload.empty?
+        # Grab the next constant name
+        constant_name = T.must(@constant_names_registered_for_autoload.shift)
```

Comment:
If you want to avoid the comments here you could extract these lines as methods:

```rb
sig { returns(String) }
def pop_next_constant_name
  T.must(@constant_names_registered_for_autoload.shift)
end

sig { params(constant_name: String).void }
def trigger_constant_autoload(constant_name)
  Reflection.constantize(constant_name, inherit: true)
end
```

Reply:
I think the code is readable enough in its current form, but I wanted to make it explicit that we were treating the array as a queue with the comments.